File: src/julia/SpaceCommsTwin/src/quantum_annealing.jl
module QuantumAnnealing

using LinearAlgebra

export IsingHamiltonian, annealing_schedule, simulate_annealing

struct IsingHamiltonian
    h::Vector{Float64}
    J::Matrix{Float64}
end

function IsingHamiltonian(n::Int)
    IsingHamiltonian(zeros(n), zeros(n, n))
end

function energy(H::IsingHamiltonian, spin_config::Vector{Int})::Float64
    E = 0.0
    for i in 1:length(H.h)
        E -= H.h[i] * spin_config[i]
    end
    for i in 1:length(H.h)
        for j in (i+1):length(H.h)
            E -= H.J[i, j] * spin_config[i] * spin_config[j]
        end
    end
    return E
end

function annealing_schedule(t::Float64, total_time::Float64)::Tuple{Float64, Float64}
    s = t / total_time
    A = 1.0 - s
    B = s
    return A, B
end

function simulate_annealing(H_final::IsingHamiltonian, n_steps::Int; T::Float64=1.0)
    n = length(H_final.h)
    spins = rand([-1, 1], n)
    energies = Float64[]

    for step in 1:n_steps
        i = rand(1:n)
        ΔE = 2.0 * H_final.h[i] * spins[i]
        for j in 1:n
            if j != i
                ΔE += 2.0 * H_final.J[i, j] * spins[i] * spins[j]
            end
        end
        if ΔE < 0.0 || rand() < exp(-ΔE / T)
            spins[i] = -spins[i]
        end
        push!(energies, energy(H_final, spins))
    end

    return spins, energies
end

end

File: src/python/space_comms_digital_twin/quantum/optimization/quantum_annealing.py
from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np


@dataclass
class AnnealingResult:
    final_state: list[int] = field(default_factory=list)
    final_energy: float = 0.0
    energy_history: list[float] = field(default_factory=list)
    n_steps: int = 0
    schedule: str = "linear"


class QuantumAnnealing:
    def __init__(self, n_qubits: int = 4):
        self.n_qubits = n_qubits

    def build_ising(self, J: np.ndarray, h: np.ndarray) -> np.ndarray:
        return J, h

    def simulate(self, J: np.ndarray, h: np.ndarray,
                 n_steps: int = 1000, temperature: float = 1.0,
                 schedule: str = "linear") -> AnnealingResult:
        spins = np.random.choice([-1, 1], self.n_qubits)
        energy_history = []

        for step in range(n_steps):
            s = step / n_steps
            if schedule == "linear":
                A, B = 1.0 - s, s
            elif schedule == "exponential":
                A = math.exp(-5 * s)
                B = 1.0 - A
            else:
                A, B = 1.0 - s, s

            current_energy = self._ising_energy(spins, J, h)

            i = np.random.randint(0, self.n_qubits)
            spins[i] *= -1
            new_energy = self._ising_energy(spins, J, h)
            delta = new_energy - current_energy

            if delta > 0 and np.random.random() > math.exp(-delta / (B * temperature)):
                spins[i] *= -1
            else:
                current_energy = new_energy

            total_energy = current_energy
            energy_history.append(float(total_energy))

        return AnnealingResult(
            final_state=list(spins),
            final_energy=energy_history[-1] if energy_history else 0.0,
            energy_history=energy_history,
            n_steps=n_steps,
            schedule=schedule,
        )

    def _ising_energy(self, spins: np.ndarray, J: np.ndarray, h: np.ndarray) -> float:
        energy = -np.dot(h, spins)
        for i in range(self.n_qubits):
            for j in range(i + 1, self.n_qubits):
                energy -= J[i, j] * spins[i] * spins[j]
        return energy

    def build_qubo(self, Q: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
        n = Q.shape[0]
        h = np.zeros(n)
        J = np.zeros((n, n))
        for i in range(n):
            h[i] = -0.5 * Q[i, i] - 0.25 * sum(Q[i, j] + Q[j, i] for j in range(n) if j != i)
            for j in range(i + 1, n):
                J[i, j] = 0.25 * (Q[i, j] + Q[j, i])
        return J, h
